README
# $Id: README 90 2011-09-02 03:54:11Z fumiyas $

Latest version may be found at https://gna.org/projects/smbldap-tools/

Bug Report:
  * Upstream: https://gna.org/bugs/?group=smbldap-tools
  * Ubuntu:   https://bugs.launchpad.net/ubuntu/+source/smbldap-tools/
  * Debian:   http://bugs.debian.org/cgi-bin/pkgreport.cgi?src=smbldap-tools

Userful documentations:
  http://www.islandlinux.org/HOWTO/openldap_samba.html (see bug in nss_ldap)


What those tools are for?
=-=-=-=-=-=-=-=-=-=-=-=-=

A collection of scripts, «over» user{add,del,mod} and group{add,del,mod}
system tools to manipulate users and groups stored in LDAP directory,
for DEN system like Samba-LDAP and pam/nss_ldap systems.

Additionnaly, some scripts are designed to ease your migration from
a Windows NT 4.0 PDC Server to a Samba-LDAP PDC Server (Killer?;-):
smbldap-populate and doc/migration_scripts/smbldap-migrate.

They are currently used with Samba 3.2 and later.

In the future, some other function may come (like backup and restore,
Novell migration tools, samba system activity report, dealing with
mail functions, compliance to RFC2307...): consult TODO.


What do SMBLDAP-TOOLS provide?
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

Those tools provide the following functions :
  * Populate LDAP database with a basic LDIF
  * Add a user or group
  * Delete a user or group
  * Modify all users or groups data (all attributes types stored in
    posixAccount, posixGroup and sambaSamAccount object class)
Taking care of :
  * Staying compatible with all standard system tools options
    (user/group{add,del,mod})
  * Be extensible for future developments
    (manipulation of shadow account options, for example)
  * Error management, in the way system tools do
Constraints :
  * Usage of Perl (portability)
  * All options must be placed in an external configuration file
  * English localization


What do SMBLDAP-TOOLS deliver?
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

Contents :
  * Scripts (see FILES)
  * User documentation in pod format, included in the sources
    (or just use the -? option)

These tools aim at delivering the same functionality as the corresponding
system tools



debian/changelog
smbldap-tools (0.9.9-1ubuntu1~aurex1) precise; urgency=low

  * Rebuild for aurex 12.04 precise

 -- Luis Antonio Garcia Gisbert <luisgg@gmail.com>  Mon, 22 Sep 2014 19:49:47 +0200

smbldap-tools (0.9.9-1ubuntu1) raring; urgency=low

  * Merge from Debian unstable (LP: #1054130). Remaining changes:
    - Apply patch from rdratlos to resolve being unable to join a Windows
      7 or Windows 2008 machine to a Samba domain due to the use of cached
      nss credentials.
    - 0020_original_doc_html_index.patch: Add index html file.

 -- Logan Rosen <logatronico@gmail.com>  Thu, 06 Dec 2012 23:27:14 -0500

smbldap-tools (0.9.9-1) unstable; urgency=low

  * New upstream release (Closes: #647860, #679935, #680939, #681350)
  * Previous upstream version closes: #670246
  * Bumped standard version (no changes needed)

 -- Sergio Talens-Oliag <sto@debian.org>  Tue, 07 Aug 2012 20:28:37 +0200

smbldap-tools (0.9.7-1ubuntu1) precise; urgency=low

  * Merge from debian unstable (LP: #889308).  Remaining changes:
    - Apply patch from rdratlos to resolve being unable to join a Windows
      7 or Windows 2008 machine to a Samba domain due to the use of cached
      nss credentials. (LP: #814898)
    - 0020_original_doc_html_index.patch: Add index html file.

 -- Leo Iannacone <l3on@ubuntu.com>  Tue, 27 Sep 2011 18:05:13 +0000

smbldap-tools (0.9.7-1) unstable; urgency=low

  * New upstream release
    (Closes: #480386, #519551, #561941, #568259, #572711, #582902).

 -- Sergio Talens-Oliag <sto@debian.org>  Tue, 27 Sep 2011 15:14:48 +0200

smbldap-tools (0.9.6-1) unstable; urgency=low

  * New upstream release
  * Bumped standard version (no changes needed)
  * Add use-Digest-SHA.patch patch to use Digest::SHA from perl core instead
    of Digest::SHA1. Drop libdigest-sha1-perl from Depends (Closes: #624065).
    Patch from Salvatore Bonaccorso <carnil@debian.org>.

 -- Sergio Talens-Oliag <sto@debian.org>  Thu, 18 Aug 2011 10:55:45 +0200

smbldap-tools (0.9.5-1ubuntu1) oneiric; urgency=low

  * Apply patch from rdratlos to resolve being unable to join a Windows
    7 or Windows 2008 machine to a Samba domain due to the use of cached
    nss credentials. (LP: #814898)

 -- Daniel T Chen <crimsun@ubuntu.com>  Mon, 01 Aug 2011 15:14:32 -0400

smbldap-tools (0.9.5-1) unstable; urgency=low

  * New upstream release (Closes: Bug#548665).

 -- Sergio Talens-Oliag <sto@debian.org>  Thu, 15 Oct 2009 21:34:50 +0200

smbldap-tools (0.9.4-1) unstable; urgency=low

  * New upstream release, development has moved to a new place:
    https://gna.org/projects/smbldap-tools/.
    The new upstream version closes some bugs:
    - smbldap-passwd (make_salt): update shadowMax (to
      $config{defaultMaxPasswordAge}) entry if script used with root
      priviledge (Closes: Bug#404109).
    - current smbldap-groupadd man page documents -o and -p options
      (Closes: Bug#358381).
  * Fix README.Debian according to Daniel Lawson's patch (Closes: #359314).
  * Added Stefan Pfetzing's patch to create a nobody user following the Debian
    style (Closes: #354092).
  * The html version of the documentation is missing from the original
    tarball, I've included a patch that provides a copy downloaded from the
    web (Closes: Bug#425402, as the updated docs are right).
  * Thanks to David Schmitt <david@schmitt.edv-bus.at> for providing a set of
    patches for the package.

 -- Sergio Talens-Oliag <sto@debian.org>  Mon, 24 Sep 2007 13:24:18 +0200

smbldap-tools (0.9.2-3) unstable; urgency=low

  * Added patch to fix UTF-8 usage on smbldap-usermod.

 -- Sergio Talens-Oliag <sto@debian.org>  Fri, 13 Jan 2006 22:02:29 +0100

smbldap-tools (0.9.2-2) unstable; urgency=low

  * Added missing depends (reported by Michael Braun).

 -- Sergio Talens-Oliag <sto@debian.org>  Sat,  7 Jan 2006 17:12:02 +0100

smbldap-tools (0.9.2-1) unstable; urgency=low

  * New upstream release (Closes: Bug#314430).

 -- Sergio Talens-Oliag <sto@debian.org>  Tue,  3 Jan 2006 19:41:59 +0100

smbldap-tools (0.9.1-2) unstable; urgency=low

  * Standars updated, no changes needed.
  * Included 'configure.pl' with the package docs (Closes: Bug#317833).

 -- Sergio Talens-Oliag <sto@debian.org>  Tue, 12 Jul 2005 13:15:08 +0200

smbldap-tools (0.9.1-1) unstable; urgency=low

  * New upstream release.
  * Fixed README.Debian; tell the user to use -u or -g options with
    smbldap-populate to change initial uid or gid (Closes: Bug#314592).

 -- Sergio Talens-Oliag <sto@debian.org>  Tue, 21 Jun 2005 09:35:55 +0200

smbldap-tools (0.9.0-1) unstable; urgency=low

  * New upstream release.
  * This version adds support to change the default HOME creation mode
    (Closes: Bug#309303).
  * Removed patch to change the configuration files paths.

 -- Sergio Talens-Oliag <sto@debian.org>  Fri, 27 May 2005 00:59:40 +0200

smbldap-tools (0.8.8-1) unstable; urgency=low

  * New upstream release (Closes: Bug#306455).

 -- Sergio Talens-Oliag <sto@debian.org>  Tue, 26 Apr 2005 20:15:52 +0200

smbldap-tools (0.8.7-4) unstable; urgency=low

  * Changed package priority to extra, as it depends on packages
    that have priority extra; I thought lintian and/or linda looked at
    those things. 

 -- Sergio Talens-Oliag <sto@debian.org>  Sun, 10 Apr 2005 12:49:00 +0200

smbldap-tools (0.8.7-3) unstable; urgency=high

  * Changed the perl dependency to a ${perl:Depends} substitution.

  * Added the dependency on libdigest-sha1-perl (Closes: Bug#298200).

 -- Sergio Talens-Oliag <sto@debian.org>  Sat,  5 Mar 2005 18:51:26 +0100

smbldap-tools (0.8.7-2) unstable; urgency=low

  * Added patch to fix configuration file locations, should have tested it
    better (Bug#296062).

 -- Sergio Talens-Oliag <sto@debian.org>  Sun, 20 Feb 2005 10:04:19 +0100

smbldap-tools (0.8.7-1) unstable; urgency=low

  * New upstream release.

 -- Sergio Talens-Oliag <sto@debian.org>  Fri, 18 Feb 2005 23:20:46 +0100

smbldap-tools (0.8.5-3-2) unstable; urgency=low


smbldap-populate.pl
#!@PERL_COMMAND@

# $Id: smbldap-populate.pl 121 2011-10-07 05:40:06Z fumiyas $

#  This code was developped by Jerome Tournier (jtournier@gmail.com) and
#  contributors (their names can be found in the CONTRIBUTORS file).

#  This was first contributed by IDEALX (http://www.opentrust.com/)

#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#  Purpose :
#       . Create an initial LDAP database suitable for Samba 3
#       . For lazy people, replace ldapadd (with only an ldif parameter)

use strict;
use warnings;
use FindBin qw($RealBin);
use smbldap_tools;
use Getopt::Std;
use Net::LDAP qw(LDAP_NO_SUCH_OBJECT);
use Net::LDAP::LDIF;
use Net::LDAP::Entry;

my %oc_by_attr = (
      "dc" =>			"dcObject",
      "o" =>			"organization",
      "ou" =>			"organizationalUnit",
      "cn" =>			"organizationalRole",
      "sambaDomainName" =>	"sambaDomain",
);

my %Options;

my $ok = getopts('a:b:e:g:i:k:l:m:r:R:u:?', \%Options);
if ( (!$ok) || ($Options{'?'}) ) {
    print_banner;
    print "Usage: $0 [-abegiklmru?] [ldif]\n";
    print "  -a user	administrator login name (default: root)\n";
    print "  -b user	guest login name (default: nobody)\n";
    print "  -e file	export ldif file\n";
    print "  -g gidNumber	first uidNumber to allocate (default: 1000)\n";
    print "  -i file	import ldif file\n";
    print "  -k uidNumber	administrator's uidNumber (default: 0)\n";
    print "  -l uidNumber	guest's uidNumber (default: 999)\n";
    print "  -m gidNumber	administrator's gidNumber (default: 0)\n";
    print "  -r ridNumber	first sambaNextRid to allocate (default: 1000)\n";
    print "  -R ridBase		sambaAlgorithmicRidBase (none)\n";
    print "  -u uidNumber	first uidNumber to allocate (default: 1000)\n";
    print "  -?		show this help message\n";

    exit (1);
}

# sanity checks
my $domain = $config{sambaDomain};
if (! defined $domain) {
    print STDERR "error: domain name not found !\n";
    print STDERR "possible reasons are:\n";
    print STDERR ". incorrect 'sambaDomain' parameter in smbldap.conf\n";
    print STDERR ". incorrect 'samba_conf' definition in smbldap_tools.pm\n";
    die;
}

my $firstuidNumber=$Options{'u'};
if (!defined($firstuidNumber)) {
    $firstuidNumber=1000;
}

my $firstgidNumber=$Options{'g'};
if (!defined($firstgidNumber)) {
    $firstgidNumber=1000;
}

my $firstridNumber=$Options{'r'};
if (!defined($firstridNumber)) {
    $firstridNumber=1000;
}

my $algorithmicRidBase = $Options{'R'};

my $adminName = $Options{'a'};
if (!defined($adminName)) {
    $adminName = "root";
}

my $guestName = $Options{'b'};
if (!defined($guestName)) {
    $guestName = "nobody";
}

my $adminUidNumber=$Options{'k'};
my $adminRid = 500;
if (!defined($adminUidNumber)) {
    $adminUidNumber = 0;
} else {
    if (defined($algorithmicRidBase)) {
	## For backward compatibility with smbldap-tools 0.9.6 and older
	$adminRid = 2 * $adminUidNumber + $algorithmicRidBase;
    }
}

my $guestUidNumber=$Options{'l'};
my $guestRid = 501;
if (!defined($guestUidNumber)) {
    $guestUidNumber = "999";
} else {
    if (defined($algorithmicRidBase)) {
	## For backward compatibility with smbldap-tools 0.9.6 and older
	$guestRid = 2 * $guestUidNumber + $algorithmicRidBase;
    }
}

my $adminGidNumber=$Options{'m'};
if (!defined($adminGidNumber)) {
    $adminGidNumber = "0";
}

print "Populating LDAP directory for domain $domain ($config{SID})\n";

my $entries_iter;

if (my $file = $Options{'i'}) {
    my $ldif = Net::LDAP::LDIF->new($file, "r", onerror => 'undef') or
	die "Cannot open file: $file: $!";
    $entries_iter = sub {
	return $ldif->read_entry;
    };
} else {
    my @entries;
    my $entry;

    print "(using builtin directory structure)\n\n";

    unless ($config{suffix} =~ /^([^=]+)=([^,]+)/) {
	die "Cannot extract first attr and value from suffix: $config{suffix}";
    }
    my $suffix_attr = $1;
    my $suffix_val = $2;
    my $suffix_oc = $oc_by_attr{$suffix_attr};
    if (!defined($suffix_oc)) {
	die "Cannot determine object class for suffix entry: $config{suffix}";
    }

    $entry = Net::LDAP::Entry->new($config{suffix},
	objectClass => $suffix_oc,
	$suffix_attr => $suffix_val,
    );
    if ($config{suffix} =~ m/(?:^|,)dc=([^,]+)/) {
	$entry->add(
	    objectClass => "organization",
	    o => $1,
	);
    }
    push(@entries, $entry);

    my @config_dn = @config{
	qw(usersdn groupsdn computersdn idmapdn sambaDomaindn sambaUnixIdPooldn)
    };
    my %entry_by_dn = ();

    for my $config_dn (@config_dn) {
	my $prefix = $config_dn;
	$prefix =~ s/,\Q$config{suffix}\E$//i;

	my $dn = $config{suffix};
	for my $node (reverse(split(/,/, $prefix))) {
	    $dn = "$node,$dn";
	    next if ($entry_by_dn{$dn});

	    unless ($node =~ /^([^=]+)=([^,]*)$/) {
		die "Cannot extract first attr and value for entry: $dn";
	    }
	    my $attr = $1;
	    my $val = $2;
	    my $oc = $oc_by_attr{$attr};
	    if (!defined($oc)) {
		die "Cannot determine object class for entry: $dn";
	    }

	    $entry = Net::LDAP::Entry->new($dn,
		objectClass => $oc,
		$attr => $val,
	    );

	    ## Add attribute required by object class
	    $entry->add(sambaSID => $config{SID}) if ($oc eq 'sambaDomain');
	    $entry->add(sn => $val) if ($oc eq 'inetOrgPerson');

	    push(@entries, $entry);
	    $entry_by_dn{$dn} = $entry;
	}
    }

    $entry = $entry_by_dn{$config{sambaDomaindn}};
    if (defined($algorithmicRidBase)) {
	$entry->add(sambaAlgorithmicRidBase => $algorithmicRidBase);
    } else {
	$entry->add(sambaNextRid => $firstridNumber);
    }

    $entry_by_dn{$config{sambaUnixIdPooldn}}->add(
	objectClass =>	"sambaUnixIdPool",
	uidNumber =>	$firstuidNumber,
	gidNumber =>	$firstgidNumber,
    );

    $entry = Net::LDAP::Entry->new("uid=$adminName,$config{usersdn}",
	objectClass =>	[qw(top person organizationalPerson inetOrgPerson sambaSAMAccount posixAccount)],
	uid =>		$adminName,
	cn =>		$adminName,
	sn =>		$adminName,
	gidNumber =>	$adminGidNumber,
	uidNumber =>	$adminUidNumber,
    );
    if ($config{shadowAccount}) {
	$entry->add(objectClass => "shadowAccount");
    }
    if (defined $config{userHome} and $config{userHome} ne "") {
	my $userHome=$config{userHome};
	$userHome=~s/\%U/$adminName/;
	$entry->add(homeDirectory => $userHome);
    } else {
	$entry->add(homeDirectory => "/nonexistent");
    }
    $entry->add(
	sambaPwdLastSet =>	0,
	sambaLogonTime =>	0,
	sambaLogoffTime =>	2147483647,
	sambaKickoffTime =>	2147483647,
	sambaPwdCanChange =>	0,
	sambaPwdMustChange =>	2147483647,
    );
    if (defined $config{userSmbHome} and $config{userSmbHome} ne "") {
	my $userSmbHome = $config{userSmbHome};
	$userSmbHome =~ s/\%U/$adminName/;
	$entry->add(sambaHomePath => $userSmbHome);
    }
    if (defined $config{userHomeDrive} and $config{userHomeDrive} ne "") {
	$entry->add(sambaHomeDrive => $config{userHomeDrive});
    }
    if (defined $config{userProfile} and $config{userProfile} ne "") {
	my $userProfile = $config{userProfile};
	$userProfile =~ s/\%U/$adminName/;
	$entry->add(sambaProfilePath => $userProfile);
    }
    $entry->add(
	sambaPrimaryGroupSID =>	"$config{SID}-512",
	sambaLMPassword =>	"XXX",
	sambaNTPassword =>	"XXX",
	sambaAcctFlags =>	"[U          ]",
	sambaSID =>		"$config{SID}-$adminRid",
	loginShell =>		"/bin/false",
	gecos =>		"Netbios Domain Administrator",
    );
    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("uid=$guestName,$config{usersdn}",
	objectClass => [qw(top person organizationalPerson inetOrgPerson sambaSAMAccount posixAccount)],
	cn =>			$guestName,
	sn =>			$guestName,
	gidNumber =>		514,
	uid =>			$guestName,
	uidNumber =>		$guestUidNumber,
	homeDirectory =>	"/nonexistent",
	sambaPwdLastSet =>	0,
	sambaLogonTime =>	0,
	sambaLogoffTime =>	2147483647,
	sambaKickoffTime =>	2147483647,
	sambaPwdCanChange =>	0,
	sambaPwdMustChange =>	2147483647,
    );
    if ($config{shadowAccount}) {
	$entry->add(objectClass => "shadowAccount");
    }
    if (defined $config{userSmbHome} and $config{userSmbHome} ne "") {
	my $userSmbHome = $config{userSmbHome};
	$userSmbHome =~ s/\%U/$guestName/;
	$entry->add(sambaHomePath => $userSmbHome);
    }
    if (defined $config{userHomeDrive} and $config{userHomeDrive} ne "") {
	$entry->add(sambaHomeDrive => $config{userHomeDrive});
    }
    if (defined $config{userProfile} and $config{userProfile} ne "") {
	my $userProfile=$config{userProfile};
	$userProfile=~s/\%U/$guestName/;
	$entry->add(sambaProfilePath => $userProfile);
    }
    $entry->add(
	sambaPrimaryGroupSID => "$config{SID}-514",
	sambaLMPassword =>	"NO PASSWORDXXXXXXXXXXXXXXXXXXXXX",
	sambaNTPassword =>	"NO PASSWORDXXXXXXXXXXXXXXXXXXXXX",
	# account disabled by default
	sambaAcctFlags =>	"[NUD        ]",
	sambaSID =>		"$config{SID}-$guestRid",
	loginShell =>		"/bin/false",
    );
    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Domain Admins,$config{groupsdn}",
	objectClass => [qw(top posixGroup sambaGroupMapping)],
	cn =>		"Domain Admins",
	gidNumber =>	512,
	memberUid =>	$adminName,
	description =>	"Netbios Domain Administrators",
	sambaSID =>	"$config{SID}-512",
	sambaGroupType =>2,
	displayName =>	"Domain Admins",
    );
    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Domain Users,$config{groupsdn}",
	objectClass => [qw(top posixGroup sambaGroupMapping)],
	cn =>		"Domain Users",
	gidNumber =>	513,
	description =>	"Netbios Domain Users",
	sambaSID =>	"$config{SID}-513",
	sambaGroupType =>2,
	displayName =>	"Domain Users",
    );
    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Domain Guests,$config{groupsdn}",
	objectClass => [qw(top posixGroup sambaGroupMapping)],
	cn =>		"Domain Guests",
	gidNumber =>	514,
	description =>	"Netbios Domain Guests Users",
	sambaSID =>	"$config{SID}-514",
	sambaGroupType =>2,
	displayName =>	"Domain Guests",
    );
    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Domain Computers,$config{groupsdn}",
	objectClass => [qw(top posixGroup sambaGroupMapping)],
	cn =>		"Domain Computers",
	gidNumber =>	515,
	description =>	"Netbios Domain Computers accounts",
	sambaSID =>	"$config{SID}-515",
	sambaGroupType =>2,
	displayName =>	"Domain Computers",
    );
    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Administrators,$config{groupsdn}",
	objectClass => [qw(top posixGroup sambaGroupMapping)],
	cn =>		"Administrators",
	gidNumber =>	544,
	description =>	"Netbios Domain Members can fully administer the computer/sambaDomainName",
	sambaSID =>	"S-1-5-32-544",
	sambaGroupType => 4,
	displayName =>	"Administrators",
    );
    push(@entries, $entry);

#    $entry = Net::LDAP::Entry->new("cn=Users,$config{groupsdn}",
#	objectClass => [qw(top posixGroup sambaGroupMapping)],
#	gidNumber =>	545,
#	cn =>		"Users",
#	description =>	"Netbios Domain Ordinary users",
#	sambaSID =>	"S-1-5-32-545",
#	sambaGroupType =>	4,
#	displayName =>	"users",
#    );
#    push(@entries, $entry);

#    $entry = Net::LDAP::Entry->new("cn=Guests,$config{groupsdn}",
#	objectClass => [qw(top posixGroup sambaGroupMapping)],
#	gidNumber =>	546,
#	cn =>		"Guests",
#	memberUid =>	$guestName,
#	description =>	"Netbios Domain Users granted guest access to the computer/sambaDomainName",
#	sambaSID =>	"S-1-5-32-546",
#	sambaGroupType =>	4,
#	displayName =>	"Guests",
#    );
#    push(@entries, $entry);

#    $entry = Net::LDAP::Entry->new("cn=Power Users,$config{groupsdn}",
#	objectClass => [qw(top posixGroup sambaGroupMapping)],
#	gidNumber =>	547,
#	cn =>		"Power Users",
#	description =>	"Netbios Domain Members can share directories and printers",
#	sambaSID =>	"S-1-5-32-547",
#	sambaGroupType =>	4,
#	displayName =>	"Power Users",
#    );
#    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Account Operators,$config{groupsdn}",
	objectClass => [qw(top posixGroup sambaGroupMapping)],
	cn =>		"Account Operators",
	gidNumber =>	548,
	description =>	"Netbios Domain Users to manipulate users accounts",
	sambaSID =>	"S-1-5-32-548",
	sambaGroupType =>	4,
	displayName =>	"Account Operators",
    );
    push(@entries, $entry);

#    $entry = Net::LDAP::Entry->new("cn=System Operators,$config{groupsdn}",
#	objectClass => [qw(top posixGroup sambaGroupMapping)],
#	gidNumber =>	549,
#	cn =>		"System Operators",
#	description =>	"Netbios Domain System Operators",
#	sambaSID =>	"S-1-5-32-549",
#	sambaGroupType =>	4,
#	displayName =>	"System Operators",
#    );
#    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Print Operators,$config{groupsdn}",
	objectClass =>	[qw(top posixGroup sambaGroupMapping)],
	cn =>		"Print Operators",
	gidNumber =>	550,
	description =>	"Netbios Domain Print Operators",
	sambaSID =>	"S-1-5-32-550",
	sambaGroupType =>	4,
	displayName =>	"Print Operators",
    );
    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Backup Operators,$config{groupsdn}",
	objectClass =>	[qw(top posixGroup sambaGroupMapping)],
	cn =>		"Backup Operators",
	gidNumber =>	551,
	description =>	"Netbios Domain Members can bypass file security to back up files",
	sambaSID =>	"S-1-5-32-551",
	sambaGroupType =>	4,
	displayName =>	"Backup Operators",
    );
    push(@entries, $entry);

    $entry = Net::LDAP::Entry->new("cn=Replicators,$config{groupsdn}",
	objectClass => [qw(top posixGroup sambaGroupMapping)],
	cn =>		"Replicators",
	gidNumber =>	552,
	description =>	"Netbios Domain Supports file replication in a sambaDomainName",
	sambaSID =>	"S-1-5-32-552",
	sambaGroupType =>	4,
	displayName =>	"Replicators",
    );
    push(@entries, $entry);

    $entries_iter = sub {
	return shift(@entries);
    };
}

if (my $file = $Options{'e'}) {
    open my $file_fh, ">$file" or die "Cannot open file: $file: $!";
    while (my $entry = $entries_iter->()) {
	$file_fh->print($entry->ldif);
    }
    print "exported ldif file: $file\n";
    exit(0);
}

my $ldap_master=connect_ldap_master();
while (my $entry = $entries_iter->()) {
    my $dn = $entry->dn;
    # we first check if the entry exist
    my $mesg = $ldap_master->search(
	base => $dn,
	scope => "base",
	filter => "objectclass=*"
    );
    if ($mesg->code && $mesg->code != LDAP_NO_SUCH_OBJECT) {
	die "failed to search entry: ", $mesg->error;
    }
    if ($mesg->count == 1) {
	print "entry $dn already exist. ";
	if ($dn eq $config{sambaUnixIdPooldn}) {
	    print "Updating it...\n";
	    my @mods;
	    foreach my $attr_tmp ($entry->attributes) {
		push(@mods,$attr_tmp=>[$entry->get_value("$attr_tmp")]);
	    }
	    my $modify = $ldap_master->modify($dn,
		'replace' => { @mods },
	    );
	    $modify->code && warn "failed to modify entry: ", $modify->error ;
	} else {
	    print "\n";
	}
    } else {
	print "adding new entry: $dn\n";
	my $result=$ldap_master->add($entry);
	$result->code && warn "failed to add entry: ", $result->error ;
    }
}
$ldap_master->unbind;

# secure the admin account
print "\nPlease provide a password for the domain $adminName: \n";
system("$RealBin/smbldap-passwd", $adminName);

exit(0);


########################################

=head1 NAME

smbldap-populate - Populate your LDAP database

=head1 SYNOPSIS

smbldap-populate [ldif-file]

=head1 DESCRIPTION

The smbldap-populate command helps to populate an LDAP server by adding the necessary entries : base suffix (doesn't abort if already there), organizational units for users, groups and computers, builtin users : Administrator and guest, builtin groups (though posixAccount only, no SambaTNG support).

-a name
Your local administrator login name (default: root)

-b name
Your local guest login name (default: nobody)

-e file
export an ldif file

-i file
import an ldif file (Options -a and -b will be ignored)

=head1 FILES

@SYSCONFDIR@/smbldap.conf : main configuration
@SYSCONFDIR@/smbldap_bind.conf : credentials for binding to the directory

=head1 SEE ALSO

smb.conf(5)

=cut

#'



# - The End
